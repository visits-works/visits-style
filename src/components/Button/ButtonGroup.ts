import React from 'react';
import styled, { StyledComponentClass, ThemeType } from '../../styled';
import Button from '.';

const ButtonGroup = styled.div`
  display: inline-block;

  &:not(:last-child) {
    margin-right: 0.5rem;
  }

  ${Button} {
    margin: 0;
    border-radius: 0;

    &:first-child {
      border-top-left-radius: 4px;
      border-bottom-left-radius: 4px;
    }

    &:last-child {
      border-top-right-radius: 4px;
      border-bottom-right-radius: 4px;
    }
  }
`;
ButtonGroup.displayName = 'ButtonGroup';

export default ButtonGroup;
