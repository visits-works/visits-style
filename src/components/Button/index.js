// @flow
import type { ComponentType } from 'react';

import { string, node } from 'prop-types';
import styled, { css } from 'styled-components';
import { darken, rgba } from 'polished';
import { findColorInvert } from '../../utils';

type Props = {
  color?: 'white' | 'light' | 'dark' | 'black' | 'primary' | 'link' | 'info' | 'success' | 'warning' | 'danger',
  size?: 'small' | 'medium' | 'large',
  outline?: boolean,
  onClick: () => void,
  theme: Theme,
}

function setColor({ theme, color, outline }: Props) {
  if (!color) {
    return css`
      background-color: ${theme.white};
      border-color: ${theme.greyLighter};
      color: ${theme.greyDarker};

      &:hover{
        border-color: ${theme.greyLight};
      }

      &:active{
        border-color: ${theme.greyDarker};
      }
    `;
  } else {
    const target = color === 'light' ? theme.greyLight : theme[color];
    const invertColor = findColorInvert(target);
    if (outline) {
      return css`
        background-color: transparent;
        border-color: ${target};
        color: ${target};

        &:hover {
          background-color: ${target};
          color: ${invertColor};
        }

        &:focus {
          box-shadow: 0 0 0 .2rem ${rgba(target, 0.5)};
        }
      `;
    } else {
      return css`
        background-color: ${target};
        border-color: transparent;
        color: ${invertColor};
        box-shadow: none;
  
        &:hover {
          background-color: ${darken(0.025, target)};
        }
  
        &:active {
          background-color: ${darken(0.05, target)};
        }

        &:focus {
          box-shadow: 0 0 0 .2rem ${rgba(target, 0.5)};
        }
      `;
    }
  }
}

const Button: ComponentType<Props>  = styled.button`
  outline: none;
  appearance: none;
  display: inline-flex;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  justify-content: center;
  border: 1px solid transparent;
  border-radius: 4px;
  height: 2.25em;
  padding: ${({ color }) => color ? '0.375em 0.75em' : 'calc(0.375em - 1px) 0.75em'};

  transition: color .15s ease-in-out, background-color .15s ease-in-out, border-color .15s ease-in-out, box-shadow .15s ease-in-out;

  ${setColor}

  &:disabled {
    pointer-events: none;
    box-shadow: none;
    opacity: 0.5;
  }
`;

Button.displayName = 'Button';

export default Button;
