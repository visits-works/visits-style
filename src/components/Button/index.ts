import * as React from 'react';
import styled, { css, StyledComponentClass } from 'styled-components';
import darken from 'polished/lib/color/darken';
import findColorInvert from '../../utils/findColorInvert';
import boxShadow from '../../utils/boxShadow';
import setSize from '../../utils/setSize';
import { ColorType, SizeType, ThemeType } from '../../types';

interface ButtonProps {
  color?: ColorType;
  size?: SizeType;
  outline?: boolean;
  onClick?: () => void;
}

function setColor({ theme, color, outline }: { theme: ThemeType, color?: ColorType, outline?: boolean }) {
  if (!color) {
    return css`
      background-color: ${theme.white};
      border-color: ${theme.border};
      color: ${theme.text};

      &:hover{
        border-color: ${theme.borderHover};
      }

      &:active{
        border-color: ${theme.borderActive};
      }
    `;
  } else if (color === 'text') {
    return css`
      background-color: transparent;
      border-color: transparent;
      color: ${theme.text};

      &:hover{
        text-decoration: underline;
      }
    `;
  }

  const target = color === 'light' ? theme.color.greyLight : theme[color];
  const invertColor = findColorInvert(target);
  if (outline) {
    return css`
      background-color: transparent;
      border-color: ${target};
      color: ${target};

      &:hover {
        background-color: ${target};
        color: ${invertColor};
      }

      &:focus {
        ${boxShadow('0.2rem', target)}
      }
    `;
  }

  return css`
    background-color: ${target};
    border-color: ${target};
    border-color: transparent;
    color: ${invertColor};
    box-shadow: none;

    &:hover {
      background-color: ${darken(0.025, target)};
    }

    &:active {
      background-color: ${darken(0.05, target)};
    }

    &:focus {
      ${boxShadow('0.2rem', target)}
    }
  `;
}

const Button = styled.button<ButtonProps>`
  position: relative;
  outline: none;
  appearance: none;
  box-sizing: border-box;
  display: inline-flex;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  justify-content: center;
  border: 1px solid transparent;
  border-radius: 4px;
  height: 2.25em;
  padding: 0.375em 0.75em;

  transition-property: background-color, color, box-shadow;
  transition-duration: 0.15s;
  transition-timing-function: ease-in-out;

  ${setColor}
  ${({ size }) => setSize('font-size', size)}

  &:disabled {
    pointer-events: none;
    box-shadow: none;
    opacity: 0.5;
  }

  &:not(:last-child) {
    margin-right: 0.5rem;
    margin-bottom: 0.5rem;
  }
`;
Button.displayName = 'Button';

export default Button;
