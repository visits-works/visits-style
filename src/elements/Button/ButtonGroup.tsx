import { type HTMLAttributes, useMemo } from 'react';
import clsx from 'clsx';

export default function ButtonGroup({ className, ...rest }: HTMLAttributes<HTMLDivElement>) {
  const name = useMemo(() => clsx(
    'flex items-center justify-center',
    '[&_button]:rounded-none [&_:first-child]:rounded-bl [&_:first-child]:rounded-tl [&_:last-child]:rounded-br [&_:last-child]:rounded-tr',
    className,
  ), [className]);
  return <div role="group" className={name} {...rest} />;
}
