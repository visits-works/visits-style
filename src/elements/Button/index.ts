import { styled,  type ExecutionContext, css } from 'styled-components';
import { HTMLAttributes } from 'react';
import { darken } from 'polished';
import findColorInvert from '../../utils/findColorInvert';
import boxShadow from '../../utils/boxShadow';
import setSize from '../../utils/setSize';
import disabledColor from '../../utils/disabledColor';
import { ColorType, ThemeType, SizeType } from '../../types';

interface ButtonSetColorConfig extends ExecutionContext {
  color?: ColorType;
  outline?: boolean;
  disabled?: boolean;
}

function setColor({ theme, color, outline, disabled }: ButtonSetColorConfig) {
  if (disabled) {
    return disabledColor(theme as ThemeType);
  }
  if (!color) {
    return css`
      border-color: ${theme.border};
      color: ${theme.text};

      &:hover {
        border-color: ${theme.borderHover};
      }

      &:active {
        border-color: ${theme.borderActive};
      }
    `;
  }
  const target = theme[color] || color;
  const invertColor = findColorInvert(theme as ThemeType, target);
  if (outline) {
    return css`
      background-color: transparent;
      border-color: ${target};
      color: ${target};

      &:hover {
        background-color: ${target};
        color: ${invertColor};
      }

      &:active {
        background-color: ${darken(0.1, target)};
      }

      &:focus:not(:active) {
        ${boxShadow('0.1rem', target, 0.5)}
      }
    `;
  }

  return css`
    background-color: ${target};
    border-color: transparent;
    color: ${invertColor};
    box-shadow: none;

    &:hover {
      color: ${invertColor};
      background-color: ${darken(0.05, target)};
    }

    &:active {
      background-color: ${darken(0.1, target)};
    }

    &:focus:not(:active) {
      ${boxShadow('0.1rem', target, 0.5)}
    }
  `;
}

export interface ButtonProps extends HTMLAttributes<HTMLButtonElement> {
  /** ボタンの色 */
  color?: ColorType;
  /** ボタンのサイズ */
  size?: SizeType;
  /** 背景が透明なボタンでする */
  outline?: boolean;
  /** 丸いボタンで表示する */
  round?: boolean;
  /** アイコンボタンとしてpaddingを同じにする */
  icon?: boolean;
}

function shouldForwardProp(name: string) {
  return (
    name !== 'color'
    && name !== 'size'
    && name !== 'outline'
    && name !== 'round'
    && name !== 'icon'
  );
}

export default styled.button.withConfig({ shouldForwardProp })<ButtonProps>`
  position: relative;
  outline: none;
  appearance: none;
  box-sizing: border-box;
  display: inline-block;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  justify-content: center;
  vertical-align: middle;
  user-select: none;
  background: transparent;
  border: 1px solid transparent;
  border-radius: ${({ round, theme }) => (round ? '2456189px' : theme.radius)};
  padding: ${({ icon }) => (icon ? '0.375em' : '0.375em 0.75em')};
  line-height: ${({ icon }) => (icon ? 1 : 1.5)};

  ${setColor}
  ${({ size }) => setSize('font-size', size)}
`;
