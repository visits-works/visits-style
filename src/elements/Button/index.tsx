import { useMemo, createElement, type ButtonHTMLAttributes } from 'react';
import clsx from 'clsx';

interface InnerButtonProps {
  variant?: 'primary' | 'info' | 'warn' | 'link' | 'outline' | 'ghost' | 'danger';
  size?: 'icon' | 'none';
}

export type ButtonProps = InnerButtonProps & ButtonHTMLAttributes<HTMLButtonElement>;

export function varientButton({ variant, size }: InnerButtonProps) {
  return clsx(
    'inline-flex items-center justify-center whitespace-nowrap transition-colors rounded-md cursor-pointer',
    'focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring',
    'disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:shrink-auto',
    {
      'bg-primary text-inverted hover:bg-primary-fore': !variant || variant === 'primary',
      'bg-info text-inverted hover:bg-info-fore': variant === 'info',
      'bg-warn text-inverted hover:bg-warn-fore': variant === 'warn',
      'bg-danger text-inverted hover:bg-danger-fore': variant === 'danger',
      'hover:bg-accent': variant === 'ghost',
      'hover:underline': variant === 'link',
      'border border-input hover:bg-accent': variant === 'outline',
    },
    {
      'px-2 py-1': !size,
      'p-1.5': size === 'icon',
    },
  );
}

export default function Button({ size, variant, className, ...rest }: ButtonProps) {
  const buttonClass = useMemo(() => (
    clsx(varientButton({ size, variant }), className)
  ), [size, variant, className]);
  return createElement('button', { className: buttonClass, ...rest });
}
