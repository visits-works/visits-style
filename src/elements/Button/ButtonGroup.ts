import { styled } from 'styled-components';
import Button from '.';

export default styled.div`
  display: inline-block;

  &:not(:last-child) {
    margin-right: 0.5rem;
  }

  ${Button} {
    margin: 0;
    border-radius: 0;

    &:hover {
      z-index: 1;
    }

    &:first-child {
      border-top-left-radius: 4px;
      border-bottom-left-radius: 4px;
    }

    &:not(:first-child) {
      margin-left: -1px;
    }

    &:last-child {
      border-top-right-radius: 4px;
      border-bottom-right-radius: 4px;
    }
  }
`;
