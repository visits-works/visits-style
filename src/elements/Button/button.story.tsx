/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import Button, { ButtonProps } from '.';
import Approved from '../Icons/Approved';
import ButtonGroup from './ButtonGroup';

const colorList = [
  '',
  'primary',
  'info',
  'link',
  'success',
  'warning',
  'danger',
  'dark',
];

const sizeList = [
  '',
  'small',
  'medium',
  'large',
];

const meta = {
  title: 'elements/Button',
  component: Button,
  tags: ['autodocs'],
  parameters: {
    control: {
      color: colorList,
      size: sizeList,
    },
  },
} satisfies Meta<ButtonProps>;

export default meta;

type Story = StoryObj<typeof meta>;

export const base: Story = {
  args: {
    children: 'Hello wolrd',
    color: '',
    size: 'medium',
    outline: false,
    round: false,
    icon: false,
  },
};

export const icon: Story = {
  render: (args) => <Button {...args}><Approved /></Button>,
  args: {
    children: 'Hello wolrd',
    color: '',
    size: 'medium',
    outline: false,
    round: false,
    icon: false,
  },
};

export const multiple: Story = {
  render: () => (
    <div>
      <Button>button 1</Button>
      <Button>ボタン 2</Button>
      <Button>BUTTON 3</Button>
    </div>
  ),
};

export const buttonGroup: Story = {
  render: () => (
    <ButtonGroup>
      <Button>button 1</Button>
      <Button>ボタン 2</Button>
      <Button>BUTTON 3</Button>
    </ButtonGroup>
  ),
};
