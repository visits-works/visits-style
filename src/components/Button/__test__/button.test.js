import React from 'react';
import { shallow } from 'enzyme';

import Button, { Wrapper } from '../index';

describe('[ Component Button ] style test', () => {
  const theme = {
    btn: {
      default: '#FFF',
    },
    text: {
      default: '#EEE',
    },
    testColor: '#DDD',
    testBg: '#000',
  };

  it('styled component render', () => {
    expect(shallow(<Wrapper theme={theme} />)).toMatchSnapshot();
  });

  it('default state', () => {
    const Component = shallow(<Wrapper theme={theme} />);
    expect(Component).toHaveStyleRule('background-color', `${theme.btn.default}`);
    expect(Component).toHaveStyleRule('color', `${theme.text.default}`);
  });

  it('change background color', () => {
    const Component = shallow(<Wrapper theme={theme} background="testBg" />);
    expect(Component).toHaveStyleRule('background-color', `${theme.testBg}`);
    expect(Component).toHaveStyleRule('color', `${theme.text.default}`);
  });

  it('change font color', () => {
    const Component = shallow(<Wrapper theme={theme} color="testColor" />);
    expect(Component).toHaveStyleRule('background-color', `${theme.btn.default}`);
    expect(Component).toHaveStyleRule('color', `${theme.testColor}`);
  });

  it('change background and font color', () => {
    const Component = shallow(<Wrapper theme={theme} background="testBg" color="testColor" />);
    expect(Component).toHaveStyleRule('background-color', `${theme.testBg}`);
    expect(Component).toHaveStyleRule('color', `${theme.testColor}`);
  });

  it('add width', () => {
    const Component = shallow(<Wrapper theme={theme} width="width-value" />);
    expect(Component).toHaveStyleRule('width', 'width-value');
  });
});

describe('[ Component Button ] component', () => {
  it('render without error', () => {
    const Component = shallow(<Button />);
    expect(Component).toMatchSnapshot();
  });

  it('onClick works properly', () => {
    const clickEvt = jest.fn();
    const Component = shallow(<Button onClick={clickEvt} />);
    Component.simulate('click');
    expect(clickEvt).toBeCalled();
  });

  it('button with Text', () => {
    const sampleText = 'Test Text';
    const Component = shallow(<Button>{sampleText}</Button>);
    expect(Component.find('div').text()).toEqual(sampleText);
  });
});
