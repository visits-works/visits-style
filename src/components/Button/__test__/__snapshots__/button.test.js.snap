// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ Component Button ] component render without error 1`] = `
<styled.button>
  <div />
</styled.button>
`;

exports[`[ Component Button ] style test styled component render 1`] = `
.c0 {
  outline: none;
  font-family: inherit;
  margin: 0;
  padding: 0;
  display: inline-block;
  font-weight: normal;
  text-align: center;
  vertical-align: middle;
  touch-action: manipulation;
  cursor: pointer;
  border: 0;
  font-size: 15px;
  border-radius: 2px;
  margin-right: 8px;
  background-color: #FFF;
  color: #EEE;
  width: auto;
  min-width: 88px;
  height: 36px;
  line-height: 36px;
}

.c0>div {
  height: auto;
  width: auto;
  padding: 0 16px;
  text-align: center;
  -webkit-transition: background-color 250ms cubic-bezier(0.4,0,0.2,1) 0ms,box-shadow 250ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4,0,0.2,1) 0ms,box-shadow 250ms cubic-bezier(0.4,0,0.2,1) 0ms;
}

.c0:disabled {
  opacity: 0.6;
  cursor: default;
}

.c0:hover:not(:disabled) {
  color: #EEE;
}

.c0:hover:not(:disabled) > div {
  background-color: rgba(255,255,255,0.18);
}

<button
  className="c0"
/>
`;
